utils/sample_qc.py

```diff
@@ -0,0 +1,487 @@
+                    median=median,
+                    mad=1.4826 * hl.median(hl.abs(elements - median))
+                ),
+                *hl.bind(
+                    lambda x: hl.tuple([x, hl.median(x)]),
+                    hl.agg.collect(ht[metric])
+                )
+            )
+        )
+
+    agg_expr = hl.struct(**{metric: get_metric_expr(ht, metric) for metric in qc_metrics})
+    if strata:
+        ht = ht.annotate_globals(metrics_stats=ht.aggregate(hl.agg.group_by(hl.tuple([ht[x] for x in strata]), agg_expr)))
+    else:
+        ht = ht.annotate_globals(metrics_stats={(): ht.aggregate(agg_expr)})
+
+    strata_exp = hl.tuple([ht[x] for x in strata]) if strata else hl.tuple([])
+
+    fail_exprs = {
+        f'fail_{metric}':
+            (ht[metric] >= ht.metrics_stats[strata_exp][metric].upper) |
+            (ht[metric] <= ht.metrics_stats[strata_exp][metric].lower)
+        for metric in qc_metrics}
+    ht = ht.transmute(**fail_exprs)
+    pop_platform_filters = make_pop_filters_expr(ht, qc_metrics)
+    return ht.annotate(pop_platform_filters=pop_platform_filters)
+
+
+def flatten_duplicate_samples_ht(dups_ht: hl.Table) -> hl.Table:
+    """
+    Flattens the result of `filter_duplicate_samples`, so that each line contains a single sample.
+    An additional annotation is added: `dup_filtered` indicating which of the duplicated samples was kept.
+
+    :param Table dups_ht: Input HT
+    :return: Flattened HT
+    :rtype: Table
+    """
+    dups_ht = dups_ht.key_by()
+    dups_ht = dups_ht.annotate(
+        dups=hl.array([(dups_ht.s_kept, True)]).extend(
+            dups_ht.filtered.map(lambda x: (x.s, False))
+        )
+    )
+    dups_ht = dups_ht.explode('dups')
+    return dups_ht.select(s=dups_ht.dups[0], dup_filtered=dups_ht.dups[1]).key_by('s')
+
+
+def add_filters_expr(
+        filters: Dict[str, hl.expr.BooleanExpression],
+        current_filters: hl.expr.SetExpression = None
+) -> hl.expr.SetExpression:
+    """
+    Creates an expression to create or add filters.
+    For each entry in the `filters` dictionary, if the value evaluates to `True`,
+    then the key is added as a filter name.
+
+    Current filters are kept if provided using `current_filters`
+
+    :param dict of str -> BooleanExpression filters: The filters and their expressions
+    :param SetExpression current_filters: The set of current filters
+    :return: An expression that can be used to annotate the filters
+    :rtype: SetExpression
+    """
+    if current_filters is None:
+        current_filters = hl.empty_set(hl.tstr)
+
+    return hl.fold(
+            lambda x, y: x.union(y),
+            current_filters,
+            [
+                hl.cond(filter_condition, hl.set([filter_name]), hl.empty_set(hl.tstr))
+                for filter_name, filter_condition in filters.items()
+            ]
+        )
+
+
+def get_platform_specific_intervals(platform_pc_loadings_ht: hl.Table, threshold: float) -> List[hl.Interval]:
```

Comment:
I'm not sure what use you envision for this function. This code currently filters the platform-PC loadings HT to those intervals where the loading weights sum (across all the PCs) to above a certain threshold. Are you sure you care about the sum, as opposed to just picking out intervals that have high loadings in a particular PC?

Reply:
Yeah, it's a good point and I'm  not sure no. I suppose some kind of weighted sum (like weighing the PC by variance explained or something) would probably make more  sense. As I mentioned in the doc, this is something experimental. I'd be totally happy to remove this if you think it's just confusing.

Reply:
@gtiao: I ended up moving this into my myoseq script for two reasons:
1) I don't  think this is ready for prime-time
2) I've now changed the way `callrate_mt` is computed so  that it's directly indexed by interval rather than by `target` (still annotating `target` in `interval_info.target`). So the function  becomes a one-liner is moot.